test/customEvents.spec.js

```diff
@@ -170,8 +170,8 @@ describe('CustomEvents', function() {
       obj2;
 
     beforeEach(function() {
-      spy = jasmine.createSpy('off');
-      spy2 = jasmine.createSpy('last');
+      spy = function() {};
+      spy2 = function() {};
```

Comment:
스파이가 아니라 빈 함수를 할당한 이유가 있나요~?

Reply:
`jest.fn()` 으로 생성되는 mock 함수는 `jasmine.createSpy()`로 생성되는 mock 함수와 다르게 `Function`을 상속받지 않습니다.
```js
// customEvents.spec.js
194  it('by handler function.', function() {
195    ce.on('play', spy);
196    ce.off(spy);
197  
198    expect(ce.events).toEqual({'play': []});
199  });

```
위 테스트의 경우 `ce.off(spy)`를 수행하면서 주어진 인자가 어떤 유형(이벤트 명 or 핸들러 함수 or 객체) 인지 확인하고 이에 따라 이벤트 등록을 해제하는 작업을 수행합니다.
이때, `spy`에 jest mock 함수를 할당하면 이를 함수가 아닌 객체로 인지하고 작업을 수행해 의도치 않은 결과가 발생합니다.

```suggestion
      var spyMock = jest.fn();
      var spyMock2 = jest.fn();
      
      spy = function() {
        spyMock();
      };
      spy2 = function() {
        spyMock2();
      };
```

위와 같이 사용하면 mock 함수도 사용하고 동작도 정상적으로 되는 것을 확인할 수 있습니다.
그러나  `spy`와 `spy2`를 핸들러 함수로 사용하는 테스트는 핸들어의 수행 여부 등을 확인하지 않고 단순히 등록, 해제가 정상적으로 이뤄지는지 확인합니다.
그래서 굳이 위 처럼 mock 함수를 사용할 필요가 없다고 판단해서 위처럼 빈 함수를 할당해서 사용했습니다.

혹시 mock 함수로 변경해서 동작하도록 수정을 할까요?

Reply:
음 뭔가 CustomEvent 내부에서 함수 여부를 체크하는 로직이 있어서 제대로 실행이 안되나 보네요. 변경된 방법이 제대로 검증할 수 없다면 유지하는게 좋을 것 같습니다.

Reply:
넵! 그러면 지금은 현재 테스트 코드를 유지하도록 하겠습니다.
추후 좀 더 좋은 방법을 찾게되면 해당 방법을 적용해보도록 하겠습니다.

Reply:
`jest.fn` 을 사용하면서 임의의 mock 함수가 함수 타입을 가지려고 한다면, 다음과 같이 `mockImplementation` API를 활용하면 될 것 같습니다.

`const spyMock = jest.fn().mockImplementation(() => {});`

실제 구동 예제는 여기서 확인해보실 수 있습니다.
https://codesandbox.io/s/objective-hill-b6j80?file=/src/index.test.js

Reply:
구동 예제의 테스트를 실제 환경에서 실행보니 마찬가지로 `Function`을 상속받지 않는 것을 확인했습니다.
또한 정확히 어떤 이유에서 다른 동작을 보이는 건지는 알 수 없습니다만, 구동 예제의 환경에서는 `jest.fn()`으로 생성되는 mock 함수도 `Function`을 상속 받는 것으로 보입니다.

```js
const spy = jest.fn();
Object.setPrototypeOf(spy, Function);

console.log(spy instanceof Function); // true
```
위 처럼 mock 함수의 프로토타입 체인에 `Function`을 강제로 추가해주면 의도한대로 동작합니다.
그러나 이렇게 mock 함수에 임의의 변형을 가하는게 괜찮은가에 대한 의문이 조금 남아있어 현재 관련 내용을 찾아보는 중 입니다!

Reply:
제 생각에 더 이상 **이 부분을 어떻게 jest에 맞게 변형해서 테스트를 통과시킬까**에 대해 시간 쏟을 필요는 없어 보여요. 이건 도구의 동작 방식일뿐이고 그를 인지 했다면, 최대한 가독성 있는 테스트로 작성할 수 있으면 될 것 같아요. 지금 테스트 코드 자체가 도구에 맞춰서 변경되다 보니 가독성이 점점 떨어지는 것 같기도 하구요. 
제가 이후에 코드를 자세히 보진 않아서 자세히는 모르겠지만, 최초 작성한 테스트코드가 더 보기 편할 것 같아요(제 예상입니다ㅎㅎ)
그리고 `ce.events`와 같은 내부 프로퍼티를 직접 검증하는 테스트가 꼭 필요한 것인지도 다시 생각해보면 좋을 것 같습니다.  그리고 얼핏 다른 테스트를 봤는데 좀 비슷한 경우가 있는 것 같아 중복되지 않는지도 한번 확인해보면 좋을 것 같아요. 
앞의 조건에 해당하지 않고 꼭 필요한 테스트라면 더 이상 jest에 맞춰 변경하지 않고 **테스트 코드의 가독성**에 맞춰 변경하거나 유지하는게 좋을 것 같습니다~